crates/turborepo-lib/src/engine/mod.rs

```diff
@@ -110,4 +114,110 @@ impl Engine<Built> {
                 .collect(),
         )
     }
+
+    pub fn validate(
+        &self,
+        package_graph: &PackageGraph,
+        concurrency: u32,
+    ) -> Result<(), Vec<ValidateError>> {
+        // TODO(olszewski) once this is hooked up to a real run, we should
+        // see if using rayon to parallelize would provide a speedup
+        let (persistent_count, mut validation_errors) = self
+            .task_graph
+            .node_indices()
+            .map(|node_index| {
+                let TaskNode::Task(task_id) = self
+                    .task_graph
+                    .node_weight(node_index)
+                    .expect("graph should contain weight for node index")
+                else {
+                    // No need to check the root node if that's where we are.
+                    return Ok(false);
+                };
+                let is_persistent = self
+                    .task_definitions
+                    .get(task_id)
+                    .map_or(false, |task_def| task_def.persistent);
+
+                for dep_index in self
+                    .task_graph
+                    .neighbors_directed(node_index, petgraph::Direction::Outgoing)
+                {
+                    let TaskNode::Task(dep_id) = self
+                        .task_graph
+                        .node_weight(dep_index)
+                        .expect("node index not in graph")
+                    else {
+                        panic!("{task_id} depends on root task node");
+                    };
+
+                    let task_definition = self.task_definitions.get(dep_id).ok_or_else(|| {
+                        ValidateError::MissingTask {
+                            task_id: dep_id.to_string(),
+                            package_name: dep_id.package().to_string(),
+                        }
+                    })?;
```

Comment:
Does it make sense to return _all_ validation issues, rather than just the first? I feel like these rules are fairly recoverable and it would help UX